File: Python3/1 - 200/43 Multiply Strings.py
# 180ms, 62.41%
class Solution:
    def multiply(self, num1, num2):
        product = [0] * (len(num1) + len(num2))
        pos = len(product)
        for n1 in reversed(num1):
            tmp = pos - 1
            for n2 in reversed(num2):
                product[tmp] += (ord(n1) - 48) * (ord(n2) - 48)
                if product[tmp] > 9:
                    product[tmp - 1] += product[tmp] // 10
                    product[tmp] %= 10
                tmp -= 1
            pos -= 1
        pos = 0
        while pos < len(product) - 1 and product[pos] == 0:
            pos += 1
        return ''.join([str(p) for p in product])[pos:]
        

File: 1 - 200/43. Multiply Strings.cc
// 12ms, 24.77%
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string multiply(string num1, string num2) {
        string sum(num1.size() + num2.size(), 0);
        for (int i = num1.size() - 1; 0 <= i; --i) {
            int carry = 0;
            for (int j = num2.size() - 1; 0 <= j; --j) {
                int tmp = sum[i + j + 1] + (num1[i] - '0') * (num2[j] - '0') + carry;
                carry = tmp / 10;
                sum[i + j + 1] = tmp - carry * 10;
            }
            sum[i] += carry;
        }

        size_t startpos = sum.find_first_not_of('\0');
        if (startpos != string::npos) {
            for (int i = startpos; i < sum.size(); ++i) sum[i] += '0';
            return sum.substr(startpos);
        }
        return "0";
    }
};
